File: src/domain/cache/expiring_value_cache.py
from __future__ import annotations
from dataclasses import dataclass, field
from threading import Lock
from typing import Callable, Generic, TypeVar
import time

T = TypeVar("T")

@dataclass
class ExpiringValueCache(Generic[T]):
    """
    Cache simples com TTL (thread-safe) para um unico valor.
    Remove o valor expirado da memoria assim que detectado.
    """

    ttl_seconds: float
    _value: T | None = None
    _ts: float | None = None
    _lock: Lock = field(default_factory=Lock)
    _hits: int = 0
    _misses: int = 0
    _evictions: int = 0

    @staticmethod
    def _now() -> float:
        return time.monotonic()

    def _is_expired(self) -> bool:
        if self._value is None or self._ts is None:
            return True
        return (self._now() - self._ts) >= self.ttl_seconds

    def _clear_unlocked(self) -> None:
        self._value = None
        self._ts = None

    def has_value(self) -> bool:
        with self._lock:
            if self._is_expired():
                if self._value is not None or self._ts is not None:
                    self._evictions += 1
                self._clear_unlocked()
                return False
            return True

    def get(self) -> T | None:
        with self._lock:
            if self._is_expired():
                if self._value is not None or self._ts is not None:
                    self._evictions += 1
                self._clear_unlocked()
                self._misses += 1
                return None
            self._hits += 1
            return self._value

    def set(self, value: T) -> None:
        with self._lock:
            self._value = value
            self._ts = self._now()

    def clear(self) -> None:
        with self._lock:
            self._clear_unlocked()

    def get_or_set(self, fetch: Callable[[], T]) -> T:
        cached = self.get()
        if cached is not None:
            return cached

        value = fetch()
        self.set(value)
        return value

    def get_stats(self) -> dict[str, int | bool]:
        with self._lock:
            return {
                "hits": self._hits,
                "misses": self._misses,
                "evictions": self._evictions,
                "has_value": self._value is not None and self._ts is not None and not self._is_expired(),
            }

File: angular-app/src/core/domain/cache/expiring-value-cache.ts
export class ExpiringValueCache<T> {
  private ttlSeconds: number;
  private value: T | null = null;
  private timestamp: number | null = null;
  private hits = 0;
  private misses = 0;
  private evictions = 0;

  constructor(ttlSeconds: number) {
    this.ttlSeconds = ttlSeconds;
  }

  private now(): number {
    return performance.now() / 1000;
  }

  private isExpired(): boolean {
    if (this.value === null || this.timestamp === null) {
      return true;
    }
    return (this.now() - this.timestamp) >= this.ttlSeconds;
  }

  private _clearInternal(): void {
    this.value = null;
    this.timestamp = null;
  }

  hasValue(): boolean {
    if (this.isExpired()) {
      if (this.value !== null || this.timestamp !== null) {
        this.evictions++;
      }
      this._clearInternal();
      return false;
    }
    return true;
  }

  get(): T | null {
    if (this.isExpired()) {
      if (this.value !== null || this.timestamp !== null) {
        this.evictions++;
      }
      this._clearInternal();
      this.misses++;
      return null;
    }
    this.hits++;
    return this.value;
  }

  set(value: T): void {
    this.value = value;
    this.timestamp = this.now();
  }

  clear(): void {
    this._clearInternal();
  }

  getOrSet(fetchFn: () => T): T {
    const cached = this.get();
    if (cached !== null) {
      return cached;
    }
    const value = fetchFn();
    this.set(value);
    return value;
  }

  getStats(): {
    hits: number;
    misses: number;
    evictions: number;
    hasValue: boolean;
  } {
    return {
      hits: this.hits,
      misses: this.misses,
      evictions: this.evictions,
      hasValue: this.value !== null && this.timestamp !== null && !this.isExpired(),
    };
  }
}
